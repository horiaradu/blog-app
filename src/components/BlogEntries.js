import React, { Component } from 'react';
import Post from './Post';
import News from './News';
import EntriesFilter from './EntriesFilter';

class BlogEntries extends Component {
  state = {
    postCheck: true,
    newsCheck: true
  };

  changePostCheck = () => {
    this.setState({ postCheck: !this.state.postCheck });
  };

  changeNewsCheck = () => {
    this.setState({ newsCheck: !this.state.newsCheck });
  };

  render() {
    const { blogs, currentUser, users } = this.props;

    return (
      <div>
        <EntriesFilter
          filter={this.state}
          changePostCheck={this.changePostCheck}
          changeNewsCheck={this.changeNewsCheck}
        />
        <div>
          {(blogs && this.state.postCheck === true) || this.state.newsCheck === true ? (
            blogs.map(blog => {
              if (blog.entry.entryType === 'post' && this.state.postCheck === true) {
                return <Post key={blog.entry.uuid} post={blog} currentUser={currentUser} users={users} />;
              } else if (blog.entry.entryType === 'news' && this.state.newsCheck === true) {
                return <News key={blog.entry.uuid} news={blog} currentUser={currentUser} users={users} />;
              } else return null;
            })
          ) : blogs && this.state.postCheck === false && this.state.newsCheck === false ? (
            <h4>You've deselected both news and posts</h4>
          ) : null}
        </div>
      </div>
    );
  }
}

export default BlogEntries;
