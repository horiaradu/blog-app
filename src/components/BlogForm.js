import React, { Component } from 'react';
import { addNewEntry, updateEntry } from '../redux/actions/blogActions';
import { connect } from 'react-redux';
import BarLevel from './BarLevel';
import '../css/blogCreator.css';
import classNames from 'classnames';

class BlogForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.entry ? props.entry.title : '',
      body: props.entry ? props.entry.body : '',
      tag: props.entry ? props.entry.tag : '',
      entryType: props.entry ? props.entry.entryType : '',
      tags: props.entry ? props.entry.tags : [],
      level: props.entry ? props.entry.level : 1,
      tagsError: false,
      entryTypeError: false,
      emptyTitleError: false,
      emptyBodyError: false,
      tagReused: false
    };
  }
  onChangeStateClick = () => {
    this.props.onUpdateClick();
  };
  onButtonUpdateEntry = () => {
    const { title, body, entryType, tags, level } = this.state;
    const entryUuid = this.props.entryUuid;
    if (entryType === '') {
      this.setState({
        entryTypeError: true
      });
    }
    if (title === '') {
      this.setState({ emptyTitleError: true });
    }
    if (body === '') {
      this.setState({ emptyBodyError: true });
    }

    if (body !== '' && title !== '' && entryType !== '') {
      const updatedPost = {
        title: title
          .slice(0, 1)
          .toUpperCase()
          .concat(title.slice(1)),
        body: body
          .slice(0, 1)
          .toUpperCase()
          .concat(body.slice(1)),
        entryType,
        tags,
        level,
        entryUuid
      };
      this.props.updateEntry(entryUuid, updatedPost);
      this.props.onUpdateClick();
    }
  };
  changeErrorState = () => {
    const { title, body } = this.state;
    if (title !== '') {
      this.setState({ emptyTitleError: false });
    }
    if (body !== '') {
      this.setState({ emptyBodyError: false });
    }
  };

  onInputChange = e => {
    this.setState({ [e.target.name]: e.target.value }, () => {
      this.changeErrorState();
    });
    if (this.state.tagReused) {
      this.setState({ tagReused: false });
    }
  };

  onRadioChange = e => {
    this.setState({
      entryType: e.target.value,
      entryTypeError: false
    });
  };

  onFormSubmit = e => {
    const { title, body, entryType, tags, level } = this.state;
    e.preventDefault();

    if (entryType === '') {
      this.setState({
        entryTypeError: true
      });
    }
    if (title === '') {
      this.setState({ emptyTitleError: true });
    }
    if (body === '') {
      this.setState({ emptyBodyError: true });
    }

    if (body !== '' && title !== '' && entryType !== '') {
      const newEntry = {
        title: title
          .slice(0, 1)
          .toUpperCase()
          .concat(title.slice(1)),
        body: body
          .slice(0, 1)
          .toUpperCase()
          .concat(body.slice(1)),
        entryType,
        tags,
        level
      };

      this.props.addNewEntry(newEntry);

      this.setState({
        title: '',
        body: '',
        entryType: '',
        tag: '',
        tags: [],
        level: 1,
        tagsError: false
      });
    }
  };

  updateLevel = level => {
    this.setState({ level: level });
  };

  onTagFormSubmit = e => {
    e.preventDefault();

    if (this.state.tags.length < 5 && !this.state.tags.includes(this.state.tag)) {
      this.setState({
        tags: this.state.tags.concat(this.state.tag),
        tag: ''
      });
    } else if (!this.state.tags.includes(this.state.tag)) {
      this.setState({ tag: '', tagsError: true });
      this.setState({ tags: this.state.tags.concat(this.state.tag), tag: '' });
    } else
      this.setState({
        tagReused: true
      });
  };

  onDeleteTag = value => {
    this.setState({
      tags: this.state.tags.filter(tag => tag !== value)
    });
    if (this.state.tags.length === 6) {
      this.setState({ tagsError: false });
    }
  };

  render() {
    const {
      title,
      body,
      tag,
      tags,
      entryType,
      tagsError,
      emptyBodyError,
      emptyTitleError,
      entryTypeError,
      tagReused
    } = this.state;

    return (
      <div>
        <div className="panelTitle">
          <h1>{this.props.entry ? 'Edit Entry' : 'Create Entry'}</h1>
          <br />
        </div>
        <div className="panelContent">
          <form onSubmit={this.onFormSubmit}>
            <div className="radioButtonsWrapper">
              <div className="radioInLine">
                <input
                  type="radio"
                  id="post"
                  name={this.state.entryType}
                  value="post"
                  checked={this.state.entryType === 'post'}
                  onChange={this.onRadioChange}
                />{' '}
                <label className="radioIcon" htmlFor="post" />
                <i className="bpost fas fa-blog">Blog</i>
              </div>

              <div className="radioInLine">
                <input
                  type="radio"
                  id="news"
                  name={this.state.entryType}
                  value="news"
                  checked={this.state.entryType === 'news'}
                  onChange={this.onRadioChange}
                />{' '}
                <label htmlFor="news" />
                <i className="bpost far fa-newspaper"> News</i>
              </div>
            </div>
            <div>
              {entryTypeError && (
                <p
                  className={classNames({
                    radioError: entryTypeError
                  })}
                >
                  Please choose a post type
                </p>
              )}
            </div>

            <div>
              <label htmlFor="title">Title:</label>
              <input
                type="text"
                id="title"
                name="title"
                placeholder="Name"
                value={title}
                onChange={this.onInputChange}
                className={classNames('inputField', {
                  inputError: emptyTitleError
                })}
              />
              {emptyTitleError && (
                <p
                  className={classNames({
                    labelError: emptyTitleError
                  })}
                >
                  Please enter a title
                </p>
              )}
            </div>

            <div>
              <label htmlFor="body">Body:</label>
              <textarea
                type="text"
                id="body"
                name="body"
                placeholder="Body"
                value={body}
                onChange={this.onInputChange}
                className={classNames('textareaField', { inputError: emptyBodyError })}
              />
              {emptyBodyError && (
                <p
                  className={classNames({
                    labelError: emptyBodyError
                  })}
                >
                  Please enter some text
                </p>
              )}
            </div>

            <div>{entryType === 'news' ? <BarLevel updateLevel={this.updateLevel} /> : null}</div>

            <div>
              <label htmlFor="tags">Tags:</label>
              <input
                className={classNames('inputField', { inputErrorTag: tagReused })}
                value={tag}
                onChange={this.onInputChange}
                onKeyDown={e => {
                  if (e.keyCode === 13) {
                    this.onTagFormSubmit(e);
                  }
                }}
                type="text"
                id="tag"
                name="tag"
                disabled={tagsError ? 'disabled' : ''}
                placeholder={tagsError ? 'Only 6 tags allowed' : 'Insert a tag'}
              />
              {this.state.tagReused && (
                <p
                  className={classNames({
                    tagReusedError: tagReused
                  })}
                >
                  Tag already used
                </p>
              )}
            </div>
            <div>
              <ul className="ulStyle">
                {tags.map(tag => (
                  <li key={tag} className="tagStyle">
                    {tag}
                    <i
                      onClick={() => this.onDeleteTag(tag)}
                      style={{ cursor: 'pointer' }}
                      className="tagsDeleteIcon fas fa-times"
                    />
                  </li>
                ))}
              </ul>
            </div>
            {this.props.entry ? null : (
              <div className="buttonWrap">
                <input className="button" type="submit" value="Save" />
              </div>
            )}
          </form>
          {this.props.entry && (
            <div className="edit-delete_wrapper">
              <span>
                <button className="cancelButton" onClick={this.onChangeStateClick}>
                  Cancel
                </button>
                <button className="updateButton" onClick={this.onButtonUpdateEntry}>
                  Update
                </button>
              </span>
            </div>
          )}
        </div>
      </div>
    );
  }
}

export default connect(
  null,
  { addNewEntry, updateEntry }
)(BlogForm);
