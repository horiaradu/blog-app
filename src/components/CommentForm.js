import React, { Component } from 'react';
import { addNewComment } from '../redux/actions/blogActions';
import { connect } from 'react-redux';
import classNames from 'classnames';

export class CommentForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      author: props.currentUser.firstName ? `${props.currentUser.firstName} ${props.currentUser.lastName}` : '',
      text: props.currentComment ? props.currentComment.text : '',
      commentDate: new Date().toLocaleString(),
      emptyAuthorError: false,
      emptyTextError: false
    };
  }

  changeErrorState = () => {
    const { author, text } = this.state;
    if (author !== '') {
      this.setState({ emptyAuthorError: false });
    }
    if (text !== '') {
      this.setState({ emptyTextError: false });
    }
  };

  onInputChange = e => {
    this.setState({ [e.target.name]: e.target.value }, () => {
      this.changeErrorState();
    });
  };

  createComment = (author, text, commentDate, userId) => {
    const newComment = {
      author: author
        .slice(0, 1)
        .toUpperCase()
        .concat(author.slice(1)),
      text: text
        .slice(0, 1)
        .toUpperCase()
        .concat(text.slice(1)),
      commentDate,
      userId: userId ? userId : ''
    };
    if (author === '') {
      this.setState({ emptyAuthorError: true });
    }
    if (text === '') {
      this.setState({ emptyTextError: true });
    }
    return newComment;
  };

  onFormSubmit = e => {
    e.preventDefault();
    const { author, text, commentDate } = this.state;
    const { userId } = this.props.currentUser;
    const newComment = this.createComment(author, text, commentDate, userId);

    if (author !== '' && text !== '') {
      this.props.addNewComment(newComment, this.props.entryUuid);
      this.setState({
        author: this.props.currentUser.firstName
          ? `${this.props.currentUser.firstName} ${this.props.currentUser.lastName}`
          : '',
        text: ''
      });
    }
  };

  onUpdateCommentClick = () => {
    const { author, text, commentDate } = this.state;
    const { userId } = this.props.currentUser;
    const updatedComment = this.createComment(author, text, commentDate, userId);
    if (author !== '' && text !== '') {
      this.props.onUpdateClick(this.props.currentComment.uuid, updatedComment, this.props.entryUuid);
    }
  };

  isEditModeOn = () => {
    return this.props.currentComment;
  };

  render() {
    const { author, text, emptyAuthorError, emptyTextError } = this.state;
    return (
      <div>
        <form id="addNewCommentForm" onSubmit={this.onFormSubmit}>
          {this.props.currentUser.firstName ? null : (
            <div>
              <label>Author:</label>
              <input
                className={classNames('inputField', { inputError: emptyAuthorError })}
                placeholder="Enter Name"
                type="text"
                name="author"
                value={author}
                onChange={this.onInputChange}
              />
              {emptyAuthorError && (
                <p
                  className={classNames({
                    labelError: emptyAuthorError
                  })}
                >
                  Please enter name
                </p>
              )}
            </div>
          )}

          <label>Comment:</label>
          <textarea
            className={classNames('textareaField', { inputError: emptyTextError })}
            placeholder="Type Something..."
            value={text}
            name="text"
            onChange={this.onInputChange}
          />
          {emptyTextError && (
            <p
              className={classNames({
                labelError: emptyAuthorError
              })}
            >
              Enter text
            </p>
          )}

          {!this.isEditModeOn() && <input type="submit" className="commentButton" value="Add Comment" />}
        </form>
        {this.isEditModeOn() && (
          <div>
            <button onClick={this.props.onCancelClick}>Cancel</button>
            <button onClick={this.onUpdateCommentClick}>Update</button>
          </div>
        )}
      </div>
    );
  }
}

export default connect(
  null,
  { addNewComment }
)(CommentForm);
