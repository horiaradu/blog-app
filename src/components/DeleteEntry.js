import React from 'react';

import { connect } from 'react-redux';
import { deleteEntry } from '../redux/actions/blogActions';

const DeleteEntry = props => {
  return (
    <div className="deleteButton">
      <i className="fas fa-trash-alt" onClick={() => props.deleteEntry(props.postUuid)} />
    </div>
  );
};

export default connect(
  null,
  { deleteEntry }
)(DeleteEntry);
