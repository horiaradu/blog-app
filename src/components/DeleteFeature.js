import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { deleteFeature } from '../redux/actions/blogActions';
import firebase from '../config/fbConfig';

const DeleteFeature = props => {
  const entryId = firebase.firestore().collection('entries');
  console.log(entryId);
  return (
    <div>
      <h4>
        <Link to="/" onClick={() => props.deleteFeature(entryId)}>
          Delete
        </Link>
      </h4>
    </div>
  );
};

export default connect(
  null,
  { deleteFeature }
)(DeleteFeature);
