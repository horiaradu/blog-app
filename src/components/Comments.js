import React, { Component } from 'react';
import CommentForm from './CommentForm';
import { deleteComment, updateComment, pinComment } from '../redux/actions/blogActions';
import { connect } from 'react-redux';
import '../css/comments.css';

class Comments extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showComments: false,
      isEditCommentModeOn: false,
      commentIdInEditMode: ''
    };
  }

  onShowClick = () => {
    this.setState({ showComments: !this.state.showComments });
  };

  onDeleteClick = (commentUuid, entryUuid, userId) => {
    this.props.deleteComment(commentUuid, entryUuid, userId);
  };

  onEditCommentClick = (e, commentUuid) => {
    if (e.target.id === commentUuid) {
      this.setState({ isEditCommentModeOn: true, commentIdInEditMode: e.target.id });
    }
  };

  onCancelClick = () => {
    this.setState({ isEditCommentModeOn: false, commentIdInEditMode: '' });
  };
  onUpdateClick = (commentUuid, data, entryUuid) => {
    this.props.updateComment(commentUuid, data, entryUuid);
    this.setState({ isEditCommentModeOn: false, commentIdInEditMode: '' });
  };
  getUserName = () => {
    const userName = this.props.users.find(user => {
      return user.userId === this.props.entryUserId;
    });
    return userName;
  };

  onPinCommentClick = (comment, entryUuid, currentUser, entryUserId) => {
    const { comments } = this.props;

    if (currentUser.userId !== entryUserId) {
      return;
    }

    const arr = comments.map(c => {
      if (c.uuid === comment.uuid) {
        return { ...c, isPinned: !comment.isPinned };
      } else return c;
    });

    const pinnedComments = arr.filter(c => {
      return c.isPinned;
    });
    const unpinnedComments = arr.filter(c => {
      return !c.isPinned;
    });

    const sortedUnpinnedComments = unpinnedComments.sort(function(x, y) {
      return new Date(x.commentDate).getTime() - new Date(y.commentDate).getTime();
    });

    const arrOfPinnedAndSortedUnpinnedComments = [...pinnedComments, ...sortedUnpinnedComments];

    this.props.pinComment(arrOfPinnedAndSortedUnpinnedComments, entryUuid);
  };

  moveElementInArray = (array, value, positionChange) => {
    const oldIndex = array.indexOf(value);
    if (oldIndex > -1) {
      let newIndex = oldIndex - positionChange;

      if (newIndex < 0) {
        newIndex = 0;
      } else if (newIndex >= array.length) {
        newIndex = array.length;
      }

      let arrayClone = array.slice();
      arrayClone.splice(oldIndex, 1);
      arrayClone.splice(newIndex, 0, value);

      return arrayClone;
    }
    return array;
  };

  movePinnedCommentUp = (comment, entryUuid) => {
    const { comments } = this.props;
    const pinnedComments = comments.filter(c => {
      return c.isPinned;
    });
    const unPinnedComments = comments.filter(c => {
      return !c.isPinned;
    });

    const sortedPinnedComments = this.moveElementInArray(pinnedComments, comment, 1);
    const newSortedArray = [...sortedPinnedComments, ...unPinnedComments];
    this.props.pinComment(newSortedArray, entryUuid);
  };

  movePinnedCommentDown = (comment, entryUuid) => {
    const { comments } = this.props;
    const pinnedComments = comments.filter(c => {
      return c.isPinned;
    });
    const unPinnedComments = comments.filter(c => {
      return !c.isPinned;
    });

    const sortedPinnedComments = this.moveElementInArray(pinnedComments, comment, -1);
    const newSortedArray = [...sortedPinnedComments, ...unPinnedComments];
    this.props.pinComment(newSortedArray, entryUuid);
  };

  hidePinDownButton = (comments, index, array) => {
    if (index === comments.length - 1 || (array[index].isPinned === true && array[index + 1].isPinned === false)) {
      return true;
    }
    return false;
  };

  render() {
    const { showComments, isEditCommentModeOn, commentIdInEditMode } = this.state;
    const { comments, currentUser, entryUuid, entryUserId } = this.props;
    const userName = this.getUserName();
    const showEditDeleteButtons = comment => {
      if (comment.userId === currentUser.userId && commentIdInEditMode !== comment.uuid) {
        return true;
      }
      return false;
    };
    const showCommentFormForCurrentCommentInEditMode = comment => {
      if (
        isEditCommentModeOn === true &&
        comment.userId === currentUser.userId &&
        comment.uuid === commentIdInEditMode
      ) {
        return true;
      }
      return false;
    };
    return (
      <div className="comments">
        <h3 className="commentHeader">
          Comments <i style={{ cursor: 'pointer' }} className="fas fa-sort-down" onClick={this.onShowClick} />
        </h3>
        {showComments &&
          comments.map((comment, index, array) => {
            return (
              <div key={comment.uuid} className="commentWrap">
                <i className="fas fa-user avatar" />
                <div className="content">
                  <h5 className={comment.isPinned ? 'pinnedAuthor' : 'author'}>{comment.author} </h5>
                  {comment.isPinned && this.getUserName() && (
                    <span className="pinnedComment pinnedText">
                      <i
                        className="fas fa-thumbtack pinnedIcon"
                        onClick={() => this.onPinCommentClick(comment, entryUuid, currentUser, entryUserId)}
                      >
                        {currentUser.userId === entryUserId && <span className="pinnedIconHoverText">Unpin</span>}
                      </i>

                      {`Pinned by ${userName.firstName} ${userName.lastName}`}
                      <span className="btn-group">
                        {index !== 0 && (
                          <i
                            onClick={() => this.movePinnedCommentUp(comment, entryUuid)}
                            className="fas fa-chevron-up pinUp"
                          />
                        )}
                        {!this.hidePinDownButton(comments, index, array) && (
                          <i
                            onClick={() => this.movePinnedCommentDown(comment, entryUuid)}
                            className="fas fa-chevron-down pinDown"
                          />
                        )}
                      </span>
                    </span>
                  )}
                  <span>
                    {currentUser.userId === entryUserId && comment.isPinned === false && (
                      <i
                        className="fas fa-thumbtack pinIcon"
                        onClick={() => this.onPinCommentClick(comment, entryUuid, currentUser, entryUserId)}
                        style={{ cursor: 'pointer' }}
                      />
                    )}
                  </span>
                  <div className="dateWrap">
                    <span className="date"> {comment.commentDate}</span>
                  </div>
                </div>

                {showCommentFormForCurrentCommentInEditMode(comment) ? (
                  <div>
                    <CommentForm
                      onCancelClick={this.onCancelClick}
                      currentUser={currentUser}
                      entryUuid={entryUuid}
                      currentComment={comment}
                      onUpdateClick={this.onUpdateClick}
                    />
                  </div>
                ) : (
                  <div className="commentBody">{comment.text}</div>
                )}
                {showEditDeleteButtons(comment) && (
                  <div className="buttonsWrap">
                    <button
                      id={comment.uuid}
                      className="editCommentButton"
                      onClick={e => this.onEditCommentClick(e, comment.uuid)}
                    >
                      Edit
                    </button>
                    <button
                      className="deleteCommentButton"
                      onClick={() => this.onDeleteClick(comment.uuid, entryUuid, currentUser.userId)}
                    >
                      Delete
                    </button>
                  </div>
                )}
              </div>
            );
          })}
        {showComments && currentUser.userId && <CommentForm entryUuid={entryUuid} currentUser={currentUser} />}
      </div>
    );
  }
}

export default connect(
  null,
  { deleteComment, updateComment, pinComment }
)(Comments);
