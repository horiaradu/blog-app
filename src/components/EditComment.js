import React, { Component } from 'react';
import CommentForm from './CommentForm';

const EditComment = props => {
  return (
    <div>
      <CommentForm
        onCancelClick={props.onCancelClick}
        currentUser={props.currentUser}
        entryUuid={props.entryUuid}
        currentComment={props.currentComment}
        onUpdateClick={props.onUpdateClick}
      />
    </div>
  );
};

export default EditComment;
