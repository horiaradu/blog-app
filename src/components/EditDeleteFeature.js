import React from 'react';
import DeleteFeature from './DeleteFeature';
const EditDeleteFeature = () => {
  return (
    <div>
      <ul>
        <li>Edit</li>
        <li>
          <DeleteFeature />
        </li>
      </ul>
    </div>
  );
};
export default EditDeleteFeature;
