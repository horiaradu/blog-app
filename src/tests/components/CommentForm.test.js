import React from 'react';
import { shallow } from 'enzyme';
import { CommentForm } from '../../components/CommentForm';
import toJSON from 'enzyme-to-json';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

let mockAddNewCommentfn;
let wrapper;

beforeEach(() => {
  mockAddNewCommentfn = jest.fn();
});

describe('Comment component', () => {
  beforeEach(() => {
    const currentUser = {
      firstName: 'Tony',
      userId: '999'
    };
    wrapper = shallow(<CommentForm addNewComment={mockAddNewCommentfn} entryUuid="123" currentUser={currentUser} />);
  });
  afterEach(() => {
    jest.restoreAllMocks();
  });
  it('should render AddNewComment correctly', () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  describe('when the form is submitted', () => {
    it('should call the mock addNewComment function', () => {
      const state = { author: 'foo', text: 'bar', commentDate: new Date(), userId: '999', isPinned: false };
      wrapper.setState(state);
      const event = {
        preventDefault() {}
      };
      wrapper.find('#addNewCommentForm').simulate('submit', event);
      expect(mockAddNewCommentfn).toHaveBeenCalledWith(
        {
          author: 'Foo',
          commentDate: state.commentDate,
          text: 'Bar',
          userId: '999',
          isPinned: false
        },
        '123'
      );
    });
    it('should reset state to default values when user is logged in', () => {
      const state = { author: 'Tony', text: 'new comment', commentDate: new Date().toLocaleString(), isPinned: false };
      const currentUser = {
        firstName: 'Tony',
        lastName: 'Stark',
        userId: '999'
      };
      wrapper = shallow(<CommentForm addNewComment={mockAddNewCommentfn} entryUuid="123" currentUser={currentUser} />);
      wrapper.setState(state);
      const event = {
        preventDefault() {}
      };
      wrapper.find('#addNewCommentForm').simulate('submit', event);

      expect(wrapper.state()).toEqual({
        author: `${currentUser.firstName} ${currentUser.lastName}`,
        commentDate: new Date().toLocaleString(),
        text: '',

        emptyAuthorError: false,
        emptyTextError: false,
        isPinned: false
      });
    });

    it('should not submit form if author is empty', () => {
      const state = {
        author: '',
        commentDate: new Date(),
        text: 'some text',
        emptyAuthorError: false,
        emptyTextError: false
      };
      wrapper.setState(state);
      const event = {
        preventDefault() {}
      };
      wrapper.find('#addNewCommentForm').simulate('submit', event);
      expect(mockAddNewCommentfn).not.toHaveBeenCalled();
    });
    it('should not submit form if text is empty', () => {
      const state = {
        author: 'author',
        commentDate: new Date(),
        text: '',
        emptyAuthorError: false,
        emptyTextError: false
      };
      wrapper.setState(state);
      const event = {
        preventDefault() {}
      };
      wrapper.find('#addNewCommentForm').simulate('submit', event);
      expect(mockAddNewCommentfn).not.toHaveBeenCalled();
    });

    it('should set emptyAuthorError to true if author is empty', () => {
      const state = {
        author: '',
        commentDate: new Date(),
        text: '',
        emptyAuthorError: false,
        emptyTextError: false
      };
      wrapper.setState(state);
      const event = {
        preventDefault() {}
      };
      wrapper.find('#addNewCommentForm').simulate('submit', event);
      expect(wrapper.state().emptyAuthorError).toBe(true);
    });
    it('should set emptyTextError to true if text is empty', () => {
      const state = {
        author: '',
        commentDate: new Date(),
        text: '',
        emptyAuthorError: false,
        emptyTextError: false
      };
      wrapper.setState(state);
      const event = {
        preventDefault() {}
      };
      wrapper.find('#addNewCommentForm').simulate('submit', event);

      expect(wrapper.state().emptyTextError).toBe(true);
    });
  });
  describe('when input is changed', () => {
    it('should update state.author when author input is changed', () => {
      const currentUser = {};
      wrapper = shallow(<CommentForm addNewComment={mockAddNewCommentfn} entryUuid="123" currentUser={currentUser} />);
      const state = {
        author: '',
        commentDate: new Date(),
        text: '',
        emptyAuthorError: false,
        emptyTextError: false
      };
      wrapper.setState(state);
      const value = 'john';
      const name = 'author';

      wrapper
        .find('input')
        .at(0)
        .simulate('change', { target: { name, value } });

      expect(wrapper.state('author')).toBe('john');
    });
    it('should update state.text when text area is changed', () => {
      const state = {
        author: '',
        commentDate: new Date(),
        text: '',
        emptyAuthorError: false,
        emptyTextError: false
      };
      wrapper.setState(state);
      const value = 'new comment';
      const name = 'text';

      wrapper
        .find('textarea')
        .at(0)
        .simulate('change', { target: { name, value } });

      expect(wrapper.state('text')).toBe('new comment');
      const mockchangeErrorState = jest.fn();
    });
  });
  describe('when changeErrorState() is called', () => {
    it('should set emptyAuthorError to false if author is not empty', () => {
      const state = {
        author: 'john',
        commentDate: new Date(),
        text: '',
        emptyAuthorError: false,
        emptyTextError: false
      };
      wrapper.setState(state);
      wrapper.instance().changeErrorState();
      expect(wrapper.state('emptyAuthorError')).toBe(false);
    });
    it('should not set emptyAuthorError to false if author is empty', () => {
      const state = {
        author: '',
        commentDate: new Date(),
        text: '',
        emptyAuthorError: true,
        emptyTextError: false
      };
      wrapper.setState(state);
      wrapper.instance().changeErrorState();
      expect(wrapper.state('emptyAuthorError')).toBe(true);
    });
    it('should set emptyTextError to false if text is not empty', () => {
      const state = {
        author: '',
        commentDate: new Date(),
        text: 'new comment',
        emptyAuthorError: false,
        emptyTextError: true
      };
      wrapper.setState(state);
      wrapper.instance().changeErrorState();
      expect(wrapper.state('emptyTextError')).toBe(false);
    });
    it('should not set emptyTextError to false if text is empty', () => {
      const state = {
        author: '',
        commentDate: new Date(),
        text: '',
        emptyAuthorError: false,
        emptyTextError: true
      };
      wrapper.setState(state);
      wrapper.instance().changeErrorState();
      expect(wrapper.state('emptyTextError')).toBe(true);
    });
  });
});
