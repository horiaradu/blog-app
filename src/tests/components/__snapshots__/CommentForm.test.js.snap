// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment component should render AddNewComment correctly 1`] = `
<div>
  <form
    id="addNewCommentForm"
    onSubmit={[Function]}
  >
    <label>
      Comment:
    </label>
    <textarea
      className="textareaField"
      name="text"
      onChange={[Function]}
      placeholder="Type Something..."
      value=""
    />
    <input
      className="commentButton"
      type="submit"
      value="Add Comment"
    />
  </form>
</div>
`;
